File: src/utils/formulaConverter.ts
/**
 * Formula Converter Utility
 * Converts LaTeX mathematical expressions to human-readable Unicode format
 */

interface SymbolMappings {
  [key: string]: string;
}

class FormulaConverter {
  private greekLetters: SymbolMappings;
  private mathSymbols: SymbolMappings;
  private functions: SymbolMappings;
  private superscripts: SymbolMappings;
  private subscripts: SymbolMappings;
  private simpleFractions: SymbolMappings;

  constructor() {
    // Greek letters mapping
    this.greekLetters = {
      'alpha': 'α', 'beta': 'β', 'gamma': 'γ', 'delta': 'δ',
      'epsilon': 'ε', 'zeta': 'ζ', 'eta': 'η', 'theta': 'θ',
      'iota': 'ι', 'kappa': 'κ', 'lambda': 'λ', 'mu': 'μ',
      'nu': 'ν', 'xi': 'ξ', 'omicron': 'ο', 'pi': 'π',
      'rho': 'ρ', 'sigma': 'σ', 'tau': 'τ', 'upsilon': 'υ',
      'phi': 'φ', 'chi': 'χ', 'psi': 'ψ', 'omega': 'ω',
      'Alpha': 'Α', 'Beta': 'Β', 'Gamma': 'Γ', 'Delta': 'Δ',
      'Epsilon': 'Ε', 'Zeta': 'Ζ', 'Eta': 'Η', 'Theta': 'Θ',
      'Iota': 'Ι', 'Kappa': 'Κ', 'Lambda': 'Λ', 'Mu': 'Μ',
      'Nu': 'Ν', 'Xi': 'Ξ', 'Omicron': 'Ο', 'Pi': 'Π',
      'Rho': 'Ρ', 'Sigma': 'Σ', 'Tau': 'Τ', 'Upsilon': 'Υ',
      'Phi': 'Φ', 'Chi': 'Χ', 'Psi': 'Ψ', 'Omega': 'Ω'
    };

    // Mathematical symbols mapping
    this.mathSymbols = {
      'infty': '∞', 'partial': '∂', 'nabla': '∇', 'sum': '∑',
      'prod': '∏', 'int': '∫', 'oint': '∮', 'pm': '±', 'mp': '∓',
      'times': '×', 'div': '÷', 'cdot': '·', 'bullet': '•',
      'cap': '∩', 'cup': '∪', 'subset': '⊂', 'supset': '⊃',
      'subseteq': '⊆', 'supseteq': '⊇', 'in': '∈', 'notin': '∉',
      'emptyset': '∅', 'forall': '∀', 'exists': '∃', 'neg': '¬',
      'land': '∧', 'lor': '∨', 'rightarrow': '→', 'leftarrow': '←',
      'leftrightarrow': '↔', 'Rightarrow': '⇒', 'Leftarrow': '⇐',
      'Leftrightarrow': '⇔', 'approx': '≈', 'neq': '≠', 'leq': '≤',
      'geq': '≥', 'll': '≪', 'gg': '≫', 'equiv': '≡', 'sim': '∼',
      'simeq': '≃', 'cong': '≅', 'propto': '∝'
    };

    // Function names mapping
    this.functions = {
      'sin': 'sin', 'cos': 'cos', 'tan': 'tan', 'cot': 'cot',
      'sec': 'sec', 'csc': 'csc', 'sinh': 'sinh', 'cosh': 'cosh',
      'tanh': 'tanh', 'log': 'log', 'ln': 'ln', 'exp': 'exp',
      'sqrt': '√', 'lim': 'lim', 'max': 'max', 'min': 'min',
      'sup': 'sup', 'inf': 'inf', 'det': 'det', 'dim': 'dim'
    };

    // Superscript mapping
    this.superscripts = {
      '0': '⁰', '1': '¹', '2': '²', '3': '³', '4': '⁴',
      '5': '⁵', '6': '⁶', '7': '⁷', '8': '⁸', '9': '⁹',
      '+': '⁺', '-': '⁻', '=': '⁼', '(': '⁽', ')': '⁾',
      'n': 'ⁿ', 'i': 'ⁱ'
    };

    // Subscript mapping
    this.subscripts = {
      '0': '₀', '1': '₁', '2': '₂', '3': '₃', '4': '₄',
      '5': '₅', '6': '₆', '7': '₇', '8': '₈', '9': '₉',
      '+': '₊', '-': '₋', '=': '₌', '(': '₍', ')': '₎',
      'a': 'ₐ', 'e': 'ₑ', 'h': 'ₕ', 'i': 'ᵢ', 'j': 'ⱼ',
      'k': 'ₖ', 'l': 'ₗ', 'm': 'ₘ', 'n': 'ₙ', 'o': 'ₒ',
      'p': 'ₚ', 'r': 'ᵣ', 's': 'ₛ', 't': 'ₜ', 'u': 'ᵤ',
      'v': 'ᵥ', 'x': 'ₓ'
    };

    // Simple fractions mapping
    this.simpleFractions = {
      '1/2': '½', '1/3': '⅓', '2/3': '⅔', '1/4': '¼',
      '3/4': '¾', '1/5': '⅕', '2/5': '⅖', '3/5': '⅗',
      '4/5': '⅘', '1/6': '⅙', '5/6': '⅚', '1/7': '⅐',
      '1/8': '⅛', '3/8': '⅜', '5/8': '⅝', '7/8': '⅞',
      '1/9': '⅑', '1/10': '⅒'
    };
  }

  /**
   * Main conversion method
   */
  convertToReadable(latex: string): string {
    if (!latex || typeof latex !== 'string') {
      return latex || '';
    }

    try {
      let result = latex.trim();
      
      // Remove LaTeX delimiters
      result = result.replace(/^\$+|\$+$/g, '');
      result = result.replace(/^\\begin\{.*?\}|\\end\{.*?\}$/g, '');
      
      // Convert fractions
      result = this.convertFractions(result);
      
      // Convert superscripts and subscripts
      result = this.convertSuperscripts(result);
      result = this.convertSubscripts(result);
      
      // Convert Greek letters
      result = this.convertGreekLetters(result);
      
      // Convert mathematical symbols
      result = this.convertMathSymbols(result);
      
      // Convert functions
      result = this.convertFunctions(result);
      
      // Convert integrals with limits
      result = this.convertIntegrals(result);
      
      // Convert summations and products
      result = this.convertSummations(result);
      
      // Convert matrices
      result = this.convertMatrices(result);
      
      // Convert binomial coefficients
      result = this.convertBinomials(result);
      
      // Clean up extra whitespace and braces
      result = this.cleanupResult(result);
      
      return result;
    } catch (error) {
      console.warn('Formula conversion error:', error);
      return latex; // Return original if conversion fails
    }
  }

  private convertFractions(text: string): string {
    // Handle simple fractions first
    for (const [fraction, unicode] of Object.entries(this.simpleFractions)) {
      const regex = new RegExp(`\\\\frac\\{${fraction.split('/')[0]}\\}\\{${fraction.split('/')[1]}\\}`, 'g');
      text = text.replace(regex, unicode);
    }
    
    // Handle general fractions
    text = text.replace(/\\frac\{([^{}]+)\}\{([^{}]+)\}/g, (match, num, den) => {
      const cleanNum = this.convertToReadable(num);
      const cleanDen = this.convertToReadable(den);
      return `(${cleanNum})/(${cleanDen})`;
    });
    
    return text;
  }

  private convertSuperscripts(text: string): string {
    return text.replace(/\^(\{([^}]+)\}|(\w))/g, (match, full, braced, single) => {
      const content = braced || single;
      return content.split('').map(char => this.superscripts[char] || char).join('');
    });
  }

  private convertSubscripts(text: string): string {
    return text.replace(/_(\{([^}]+)\}|(\w))/g, (match, full, braced, single) => {
      const content = braced || single;
      return content.split('').map(char => this.subscripts[char] || char).join('');
    });
  }

  private convertGreekLetters(text: string): string {
    for (const [latex, unicode] of Object.entries(this.greekLetters)) {
      const regex = new RegExp(`\\\\${latex}\\b`, 'g');
      text = text.replace(regex, unicode);
    }
    return text;
  }

  private convertMathSymbols(text: string): string {
    for (const [latex, unicode] of Object.entries(this.mathSymbols)) {
      const regex = new RegExp(`\\\\${latex}\\b`, 'g');
      text = text.replace(regex, unicode);
    }
    return text;
  }

  private convertFunctions(text: string): string {
    for (const [latex, readable] of Object.entries(this.functions)) {
      const regex = new RegExp(`\\\\${latex}\\b`, 'g');
      text = text.replace(regex, readable);
    }
    return text;
  }

  private convertIntegrals(text: string): string {
    // Integral with limits
    text = text.replace(/\\int_\{([^}]+)\}\^\{([^}]+)\}/g, '∫[from $1 to $2]');
    text = text.replace(/\\int_([a-zA-Z0-9])\^([a-zA-Z0-9])/g, '∫[from $1 to $2]');
    text = text.replace(/\\int/g, '∫');
    
    return text;
  }

  private convertSummations(text: string): string {
    // Summation with limits
    text = text.replace(/\\sum_\{([^}]+)\}\^\{([^}]+)\}/g, '∑[from $1 to $2]');
    text = text.replace(/\\sum_([a-zA-Z0-9=]+)\^([a-zA-Z0-9∞]+)/g, '∑[from $1 to $2]');
    text = text.replace(/\\sum/g, '∑');
    
    // Product with limits
    text = text.replace(/\\prod_\{([^}]+)\}\^\{([^}]+)\}/g, '∏[from $1 to $2]');
    text = text.replace(/\\prod_([a-zA-Z0-9=]+)\^([a-zA-Z0-9∞]+)/g, '∏[from $1 to $2]');
    text = text.replace(/\\prod/g, '∏');
    
    return text;
  }

  private convertMatrices(text: string): string {
    // Simple matrix conversion
    text = text.replace(/\\begin\{matrix\}(.*?)\\end\{matrix\}/gs, (match, content) => {
      const rows = content.split('\\\\').map((row: string) => 
        row.trim().split('&').map((cell: string) => cell.trim()).join(' ')
      );
      return `[${rows.join('; ')}]`;
    });
    
    return text;
  }

  private convertBinomials(text: string): string {
    text = text.replace(/\\binom\{([^}]+)\}\{([^}]+)\}/g, 'C($1,$2)');
    text = text.replace(/\\choose/g, 'C');
    
    return text;
  }

  private cleanupResult(text: string): string {
    // Remove extra braces
    text = text.replace(/\{([^{}]*)\}/g, '$1');
    
    // Remove backslashes from remaining LaTeX commands
    text = text.replace(/\\([a-zA-Z]+)/g, '$1');
    
    // Clean up multiple spaces
    text = text.replace(/\s+/g, ' ');
    
    // Remove leading/trailing whitespace
    text = text.trim();
    
    return text;
  }
}

// Create and export a singleton instance
const formulaConverter = new FormulaConverter();

export default formulaConverter;

File: app/utils/formula_converter.py
#!/usr/bin/env python
"""
LaTeX to human-readable formula converter
"""

import re
from typing import Dict, List

class FormulaConverter:
    """Convert LaTeX formulas to human-readable format"""
    
    def __init__(self):
        # Greek letters mapping
        self.greek_letters = {
            r'\\alpha': 'α', r'\\beta': 'β', r'\\gamma': 'γ', r'\\delta': 'δ',
            r'\\epsilon': 'ε', r'\\zeta': 'ζ', r'\\eta': 'η', r'\\theta': 'θ',
            r'\\iota': 'ι', r'\\kappa': 'κ', r'\\lambda': 'λ', r'\\mu': 'μ',
            r'\\nu': 'ν', r'\\xi': 'ξ', r'\\omicron': 'ο', r'\\pi': 'π',
            r'\\rho': 'ρ', r'\\sigma': 'σ', r'\\tau': 'τ', r'\\upsilon': 'υ',
            r'\\phi': 'φ', r'\\chi': 'χ', r'\\psi': 'ψ', r'\\omega': 'ω',
            # Capital Greek letters
            r'\\Alpha': 'Α', r'\\Beta': 'Β', r'\\Gamma': 'Γ', r'\\Delta': 'Δ',
            r'\\Epsilon': 'Ε', r'\\Zeta': 'Ζ', r'\\Eta': 'Η', r'\\Theta': 'Θ',
            r'\\Iota': 'Ι', r'\\Kappa': 'Κ', r'\\Lambda': 'Λ', r'\\Mu': 'Μ',
            r'\\Nu': 'Ν', r'\\Xi': 'Ξ', r'\\Omicron': 'Ο', r'\\Pi': 'Π',
            r'\\Rho': 'Ρ', r'\\Sigma': 'Σ', r'\\Tau': 'Τ', r'\\Upsilon': 'Υ',
            r'\\Phi': 'Φ', r'\\Chi': 'Χ', r'\\Psi': 'Ψ', r'\\Omega': 'Ω'
        }
        
        # Mathematical symbols
        self.math_symbols = {
            r'\\infty': '∞', r'\\partial': '∂', r'\\nabla': '∇',
            r'\\pm': '±', r'\\mp': '∓', r'\\times': '×', r'\\div': '÷',
            r'\\cdot': '·', r'\\bullet': '•', r'\\circ': '∘',
            r'\\leq': '≤', r'\\geq': '≥', r'\\neq': '≠', r'\\approx': '≈',
            r'\\equiv': '≡', r'\\propto': '∝', r'\\sim': '∼',
            r'\\in': '∈', r'\\notin': '∉', r'\\subset': '⊂', r'\\supset': '⊃',
            r'\\subseteq': '⊆', r'\\supseteq': '⊇', r'\\cup': '∪', r'\\cap': '∩',
            r'\\emptyset': '∅', r'\\forall': '∀', r'\\exists': '∃',
            r'\\rightarrow': '→', r'\\leftarrow': '←', r'\\leftrightarrow': '↔',
            r'\\Rightarrow': '⇒', r'\\Leftarrow': '⇐', r'\\Leftrightarrow': '⇔'
        }
        
        # Function names
        self.functions = {
            r'\\sin': 'sin', r'\\cos': 'cos', r'\\tan': 'tan',
            r'\\sec': 'sec', r'\\csc': 'csc', r'\\cot': 'cot',
            r'\\sinh': 'sinh', r'\\cosh': 'cosh', r'\\tanh': 'tanh',
            r'\\log': 'log', r'\\ln': 'ln', r'\\exp': 'exp',
            r'\\sqrt': '√', r'\\lim': 'lim', r'\\max': 'max', r'\\min': 'min'
        }
        
        # Unicode fractions for simple cases
        self.simple_fractions = {
            '1/2': '½', '1/3': '⅓', '2/3': '⅔', '1/4': '¼', '3/4': '¾',
            '1/5': '⅕', '2/5': '⅖', '3/5': '⅗', '4/5': '⅘', '1/6': '⅙',
            '5/6': '⅚', '1/7': '⅐', '1/8': '⅛', '3/8': '⅜', '5/8': '⅝',
            '7/8': '⅞', '1/9': '⅑', '1/10': '⅒'
        }
        
        # Superscript and subscript mappings
        self.superscript_map = str.maketrans('0123456789+-=()abcdefghijklmnopqrstuvwxyz',
                                           '⁰¹²³⁴⁵⁶⁷⁸⁹⁺⁻⁼⁽⁾ᵃᵇᶜᵈᵉᶠᵍʰⁱʲᵏˡᵐⁿᵒᵖᵠʳˢᵗᵘᵛʷˣʸᶻ')
        self.subscript_map = str.maketrans('0123456789+-=()abcdefghijklmnopqrstuvwxyz',
                                         '₀₁₂₃₄₅₆₇₈₉₊₋₌₍₎ₐᵦᶜᵈₑᶠᵍₕᵢⱼₖₗₘₙₒₚᵩᵣₛₜᵤᵥwₓᵧᵤ')
    
    def convert_to_readable(self, latex_formula: str) -> str:
        """Convert LaTeX formula to human-readable format"""
        if not latex_formula or not latex_formula.strip():
            return latex_formula
            
        formula = latex_formula.strip()
        
        try:
            # Remove common LaTeX delimiters
            formula = re.sub(r'^\$+|\$+$', '', formula)
            formula = re.sub(r'^\\begin\{equation\}|\\end\{equation\}$', '', formula)
            formula = re.sub(r'^\\begin\{align\}|\\end\{align\}$', '', formula)
            
            # Convert fractions
            formula = self._convert_fractions(formula)
            
            # Convert integrals, sums, products
            formula = self._convert_integrals(formula)
            formula = self._convert_sums(formula)
            formula = self._convert_products(formula)
            
            # Convert matrices
            formula = self._convert_matrices(formula)
            
            # Convert binomial coefficients
            formula = self._convert_binomials(formula)
            
            # Convert superscripts and subscripts
            formula = self._convert_superscripts(formula)
            formula = self._convert_subscripts(formula)
            
            # Convert Greek letters
            for latex, unicode_char in self.greek_letters.items():
                formula = re.sub(latex + r'\b', unicode_char, formula)
            
            # Convert mathematical symbols
            for latex, unicode_char in self.math_symbols.items():
                formula = formula.replace(latex, unicode_char)
            
            # Convert function names
            for latex, readable in self.functions.items():
                formula = re.sub(latex + r'\b', readable, formula)
            
            # Convert delimiters
            formula = self._convert_delimiters(formula)
            
            # Clean up extra spaces and formatting
            formula = self._clean_formula(formula)
            
            return formula
            
        except Exception as e:
            print(f"Error converting formula: {e}")
            return latex_formula
    
    def _convert_fractions(self, formula: str) -> str:
        """Convert LaTeX fractions to readable format"""
        # Handle simple fractions first
        for frac, unicode_frac in self.simple_fractions.items():
            pattern = r'\\frac\{' + re.escape(frac.split('/')[0]) + r'\}\{' + re.escape(frac.split('/')[1]) + r'\}'
            formula = re.sub(pattern, unicode_frac, formula)
        
        # Handle complex fractions
        def replace_frac(match):
            numerator = match.group(1)
            denominator = match.group(2)
            return f"({numerator})/({denominator})"
        
        formula = re.sub(r'\\frac\{([^{}]*(?:\{[^{}]*\}[^{}]*)*)\}\{([^{}]*(?:\{[^{}]*\}[^{}]*)*)\}', 
                        replace_frac, formula)
        
        return formula
    
    def _convert_integrals(self, formula: str) -> str:
        """Convert integrals to readable format"""
        # Definite integrals with limits
        formula = re.sub(r'\\int_\{([^}]+)\}\^\{([^}]+)\}', r'∫[from \1 to \2]', formula)
        # Indefinite integrals
        formula = re.sub(r'\\int', '∫', formula)
        return formula
    
    def _convert_sums(self, formula: str) -> str:
        """Convert summations to readable format"""
        formula = re.sub(r'\\sum_\{([^}]+)\}\^\{([^}]+)\}', r'Σ[from \1 to \2]', formula)
        formula = re.sub(r'\\sum', 'Σ', formula)
        return formula
    
    def _convert_products(self, formula: str) -> str:
        """Convert products to readable format"""
        formula = re.sub(r'\\prod_\{([^}]+)\}\^\{([^}]+)\}', r'Π[from \1 to \2]', formula)
        formula = re.sub(r'\\prod', 'Π', formula)
        return formula
    
    def _convert_matrices(self, formula: str) -> str:
        """Convert matrices to readable format"""
        # Simple matrix conversion
        formula = re.sub(r'\\begin\{matrix\}(.*?)\\end\{matrix\}', 
                        lambda m: '[' + m.group(1).replace('\\\\', '; ').replace('&', ', ') + ']', 
                        formula, flags=re.DOTALL)
        return formula
    
    def _convert_binomials(self, formula: str) -> str:
        """Convert binomial coefficients"""
        formula = re.sub(r'\\binom\{([^}]+)\}\{([^}]+)\}', r'C(\1,\2)', formula)
        return formula
    
    def _convert_superscripts(self, formula: str) -> str:
        """Convert superscripts to Unicode"""
        def replace_sup(match):
            content = match.group(1)
            if len(content) == 1 and content in '0123456789+-=()abcdefghijklmnopqrstuvwxyz':
                return content.translate(self.superscript_map)
            else:
                return f"^({content})"
        
        formula = re.sub(r'\^{([^}]+)}', replace_sup, formula)
        formula = re.sub(r'\^([0-9a-zA-Z])', lambda m: m.group(1).translate(self.superscript_map), formula)
        return formula
    
    def _convert_subscripts(self, formula: str) -> str:
        """Convert subscripts to Unicode"""
        def replace_sub(match):
            content = match.group(1)
            if len(content) == 1 and content in '0123456789+-=()abcdefghijklmnopqrstuvwxyz':
                return content.translate(self.subscript_map)
            else:
                return f"_({content})"
        
        formula = re.sub(r'_{([^}]+)}', replace_sub, formula)
        formula = re.sub(r'_([0-9a-zA-Z])', lambda m: m.group(1).translate(self.subscript_map), formula)
        return formula
    
    def _convert_delimiters(self, formula: str) -> str:
        """Convert LaTeX delimiters"""
        formula = re.sub(r'\\left\(', '(', formula)
        formula = re.sub(r'\\right\)', ')', formula)
        formula = re.sub(r'\\left\[', '[', formula)
        formula = re.sub(r'\\right\]', ']', formula)
        formula = re.sub(r'\\left\{', '{', formula)
        formula = re.sub(r'\\right\}', '}', formula)
        formula = re.sub(r'\\left\|', '|', formula)
        formula = re.sub(r'\\right\|', '|', formula)
        return formula
    
    def _clean_formula(self, formula: str) -> str:
        """Clean up the formula"""
        # Remove extra spaces
        formula = re.sub(r'\s+', ' ', formula)
        # Remove empty braces
        formula = re.sub(r'\{\}', '', formula)
        # Clean up spacing around operators
        formula = re.sub(r'\s*([+\-*/=<>])\s*', r' \1 ', formula)
        return formula.strip()

# Global instance
formula_converter = FormulaConverter()
